ldms/src/sampler/llnl/lustre_client.c

```diff
@@ -0,0 +1,641 @@
+        rc = ldms_schema_meta_array_add(sch, "fs_name", LDMS_V_CHAR_ARRAY, 64);
+        if (rc < 0)
+                goto err2;
+        rc = ldms_schema_meta_array_add(sch, "llite", LDMS_V_CHAR_ARRAY, 64);
+        if (rc < 0)
+                goto err2;
+	rc = ldms_schema_meta_add(sch, "component_id", LDMS_V_U64);
+        if (rc < 0)
+                goto err2;
+	job_id_index = LJI_ADD_JOBID(sch);
+	if (job_id_index < 0) {
+		goto err2;
+	}
+        /* add llite stats entries */
+        for (i = 0; llite_stats_uint64_t_entries[i] != NULL; i++) {
+                rc = ldms_schema_metric_add(sch, llite_stats_uint64_t_entries[i],
+                                            LDMS_V_U64);
+                if (rc < 0)
+                        goto err2;
+                rc = metric_lookup_tree_insert(llite_stats_uint64_t_entries[i], rc);
+                if (rc != 0) {
+                        goto err2;
+                }
+        }
+
+        llite_general_schema = sch;
+
+        return 0;
+err2:
+        ldms_schema_delete(sch);
+err1:
+        log_fn(LDMSD_LERROR, SAMP" lustre_client schema creation failed\n");
+        return -1;
+}
+
+static void llite_general_schema_fini()
+{
+        log_fn(LDMSD_LDEBUG, SAMP" llite_general_schema_fini()\n");
+        if (llite_general_schema != NULL) {
+                ldms_schema_delete(llite_general_schema);
+                llite_general_schema = NULL;
+                metric_lookup_tree_destroy();
+        }
+}
+
+static void llite_general_destroy(ldms_set_t set)
+{
+        /* ldms_set_unpublish(set); */
+        ldms_set_delete(set);
+}
+
+/* must be schema created by llite_general_schema_create() */
+static ldms_set_t llite_general_create(const char *host_name, const char *producer_name,
+                                       const char *fs_name, const char *llite_name,
+                                       uint64_t component_id)
+{
+        ldms_set_t set;
+        int index;
+        char instance_name[256];
+
+        log_fn(LDMSD_LDEBUG, SAMP" llite_general_create()\n");
+        snprintf(instance_name, sizeof(instance_name), "%s/%s",
+                 producer_name, llite_name);
+        set = ldms_set_new(instance_name, llite_general_schema);
+        ldms_set_producer_name_set(set, producer_name);
+        index = ldms_metric_by_name(set, "hostname");
+        ldms_metric_array_set_str(set, index, host_name);
+        index = ldms_metric_by_name(set, "fs_name");
+        ldms_metric_array_set_str(set, index, fs_name);
+        index = ldms_metric_by_name(set, "llite");
+        ldms_metric_array_set_str(set, index, llite_name);
+        index = ldms_metric_by_name(set, "component_id");
+        ldms_metric_set_u64(set, index, component_id);
+        /* ldms_set_publish(set); */
+
+        return set;
+}
+
+static void llite_stats_sample(const char *stats_path,
+                                   ldms_set_t general_metric_set)
+{
+        FILE *sf;
+        char buf[512];
```

Comment:
If you can put a reasonable upper bound on the size of the 'stats' file (4k seems likely), it could be faster to read the entire file in a single read rather than multiple fgets queries. The kernel security features of late to mitigate TLB attacks added overhead. Unfortunately, there is no convenient string equivalent of fgets, so I can live with this version going to production and the improvement being done whenever we have a global solution for all our samplers that use fgets.

Reply:
fgets() doesn't necessarily imply an underlying read() each time it is used.  It depends on the buffering method selected.  Perhaps one could use setvbuf() if this became enough of a concern?